marimo/_plugins/ui/_impl/chat/chat.py

```diff
@@ -0,0 +1,175 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, Final, List, Optional
+
+from marimo._output.formatting import as_html
+from marimo._output.rich_help import mddoc
+from marimo._plugins.ui._core.ui_element import UIElement
+from marimo._plugins.ui._impl.chat.convert import model_from_callable
+from marimo._plugins.ui._impl.chat.types import (
+    ChatMessage,
+    ChatModel,
+    ChatModelConfig,
+    from_chat_message_dict,
+)
+from marimo._runtime.functions import EmptyArgs, Function
+
+
+@dataclass
+class SendMessageRequest:
+    messages: List[ChatMessage]
+    config: ChatModelConfig
+
+
+@dataclass
+class GetChatHistoryResponse:
+    messages: List[ChatMessage]
+
+
+DEFAULT_SYSTEM_MESSAGE = (
+    "You are a helpful assistant specializing in data science."
+)
+
+
+@mddoc
+class chat(UIElement[Dict[str, Any], List[ChatMessage]]):
+    """
+    A chatbot UI element for interactive conversations.
+
+    **Example - Using a custom model.**
+
+    You can define a custom chat model that takes a
+    prompt, previous messages, and optional attachments as input
+    and returns a response.
+
+    The response can be an object, a marimo UI element, or a string.
+
+    ```python
+    def my_rag_model(prompt, messages, attachments):
+        docs = find_docs(prompt)
+        prompt = template(prompt, docs, messages)
+        response = query(prompt)
+        if is_dataset(response):
+            return dataset_to_chart(response)
+        return response
+
+
+    chat = mo.ai.chatbot(
+        my_rag_model,
+        system_message="You are a helpful assistant.",
```

Comment:
Is `system_message` used then the "model" is a callable? It didn't look like it from the code